### PR TITLE
Add a check for linux version consistency

### PR DESCRIPTION
**Pull request by yorek**
This fix makes sure that the linux version name stored in apt sources .list file is coherent with the os reported value. If this is not true, in fact, installing extensions my fail with error

`ImportError: libffi.so.5: cannot open shared object file: No such file or directory`

that is a little obscure. This PR make sure that the error is more explicit and helpful, saving hours of frustration to each one who wants to use extensions.

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5675
This is an **experimental preview** for @microsoft users.

**Comment by derekbekoe** [comment]
Hi @yorek 
Thanks for the PR and I understand the frustration.
However, I want to avoid platform specific code like this in the CLI. Also, there is no guarantee that the filename will be as in the PR.

I would prefer adding a 'Troubleshooting' section to https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions that documents this and linking to this documentation in the [error message](https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/custom.py#L155) 'See <url> for common troubleshooting'.

@sptramer What do you think about adding such a section to the extension doc.
In short, the wrong distro package was installed.
Long version, https://github.com/Azure/azure-cli-extensions/issues/65#issuecomment-366270381

**Comment by sptramer** [comment]
If we have common problems that are seen during extension install, and you can provide steps for remediation, then I think that having a "troubleshooting" section is fine as long as it can be kept reasonably up to date. Engineering may be responsible for maintaining the content so that they can respond to issues like Azure/azure-cli-extensions#65 but I can provide some copyediting.

**Comment by yorek** [comment]
Hi @derekbekoe, I understand, and surely a troubleshooting page it is a viable alternative, but if I'd still think it is way better if the application can check for any know problems and help the user to have them fixed. The application should simply detect the problem and then redirect the user to the troubleshooting page: that would be a great user experience, instead of giving the error and than just let the user figure out how to fix it.

Let my try to address your totally legitimate concerns:

Platform specific code is already everywhere in AZ CLI: I've checked the code and there are really many places (literally, dozens) where platform-specific checks are already done, so it is not that this new one will be the first or the only one.

The filename is the same filename used in the installation guide, and is the one expected by apt, so unless apt and/or the installation guide are changed it should be pretty stable. In addition to that, since other distro may not use APT get, the code handles that situation by just ignoring any error that may arise if the file is not there, so being totally transparent to different distros

I really hope this points are enough to convince you that AZ CLI should have this PR included, since it is a very small one but that may have a big impact in everyone who may face the problem.

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 1]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
...
+                    logger.error('An error occurred while checking linux distribution version source list consistency.')
+                    logger.error(err)                    
+
+                if (current_linux_dist_name != stored_linux_dist_name):
+                    raise CLIError("mismatch distribution name in /etc/apt/sources.list.d/azure-cli.list file")                    
+
             if (namespace.extension_name and namespace.source) or (not namespace.extension_name and not namespace.source):
                 raise CLIError("usage error: --name NAME | --source SOURCE")
 
+        #raise CLIError("installation error: .list file contains does not point to the correct distrubtion")
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Please remove commented code.

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 2]
```diff
@@ -3,17 +3,24 @@
...
 from knack.util import CLIError
+from knack.log import get_logger
 
 from azure.cli.core import AzCommandsLoader
 from azure.cli.core.commands import CliCommandType
 
 import azure.cli.command_modules.extension._help  # pylint: disable=unused-import
 
+IS_WINDOWS = sys.platform.lower() in ['windows', 'win32']
+LIST_FILE_PATH = '/etc/apt/sources.list.d/azure-cli.list'
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Use `os.path.join('etc', 'apt', etc.)` instead.
Don't want to hardcode `/`

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 3]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
...
+                    logger.debug('Reported by API: %s', current_linux_dist_name)
+                except Exception as err:
+                    current_linux_dist_name = ""
+                    stored_linux_dist_name = ""
+                    logger.error('An error occurred while checking linux distribution version source list consistency.')
+                    logger.error(err)                    
+
+                if (current_linux_dist_name != stored_linux_dist_name):
+                    raise CLIError("mismatch distribution name in /etc/apt/sources.list.d/azure-cli.list file")                    
+
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Instead of here, can you put the code right before we call pip?
https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/custom.py#L139

That way it would handle both add and upgrades of extensions.

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 4]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
             return [OrderedDict([('Name', r)]) for r in results]
 
         def validate_extension_add(namespace):
+            if IS_WINDOWS == False:
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: prefer: `if not IS_WINDOWS`

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 5]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
...
+                try:
+                    logger.debug('Reading from: %s', LIST_FILE_PATH)
+                    list_file = open(LIST_FILE_PATH, 'r')
+                    package_source = list_file.read() 
+                    list_file.close()
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Use:
```
with open(LIST_FILE_PATH, 'r'):
   package_source = list_file.read() 
   stored_linux_dist_name = package_source.split(" ")[3]
   ...
```

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 6]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
             return [OrderedDict([('Name', r)]) for r in results]
 
         def validate_extension_add(namespace):
+            if IS_WINDOWS == False:
+                try:
+                    logger.debug('Reading from: %s', LIST_FILE_PATH)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Prefix all log messages with 'Linux distro check: ' so we know where the logs are coming from.
Thanks for using the logger though!

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 7]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
...
+                    stored_linux_dist_name = package_source.split(" ")[3]
+                    logger.debug('Found in list file: %s', stored_linux_dist_name)
+
+                    current_linux_dist_name = platform.linux_distribution()[2]
+                    logger.debug('Reported by API: %s', current_linux_dist_name)
+                except Exception as err:
+                    current_linux_dist_name = ""
+                    stored_linux_dist_name = ""
+                    logger.error('An error occurred while checking linux distribution version source list consistency.')
+                    logger.error(err)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Prefer: `logger.debug`.
The user can't really do much if an error happens here so would rather not show anything.

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py [hunk 8]
```diff
@@ -30,9 +37,32 @@ def transform_extension_list_available(results):
...
+                    current_linux_dist_name = platform.linux_distribution()[2]
+                    logger.debug('Reported by API: %s', current_linux_dist_name)
+                except Exception as err:
+                    current_linux_dist_name = ""
+                    stored_linux_dist_name = ""
+                    logger.error('An error occurred while checking linux distribution version source list consistency.')
+                    logger.error(err)                    
+
+                if (current_linux_dist_name != stored_linux_dist_name):
+                    raise CLIError("mismatch distribution name in /etc/apt/sources.list.d/azure-cli.list file")
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: Prefer: `logger.warning` instead of stopping the execution of the whole command. Also, consider adding an actionable message. e.g. "If command fails, install the appropriate package for your distribution. /etc/apt/sources.list.d/azure-cli.list has artful, current distro is trusty" or something like that.

There could be edge cases where the user actually wants a different distribution.
e.g. We don't support Ubuntu Bionic yet so if a user installed the Artful package, everything 'should' still be okay so we don't want to prevent them from executing the command.

**Comment by yorek** [comment]
Great, will do!

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 1 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 2 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 3 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 4 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 5 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 6 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 7 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

File: src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py (hunk 8 above)
Comment (yorek) [reply src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/__init__.py]: done

**Comment by yorek** [comment]
I made the changes and create a new PR, closing this one.
